rust/sql/src/postgres/reader.rs

```diff
@@ -0,0 +1,894 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::convert::TryFrom;
+use std::{
+    io::{Read, Write},
+    sync::Arc,
+};
+
+use arrow::array::*;
+use arrow::buffer::Buffer;
+use arrow::datatypes::{
+    DataType, DateUnit, Field, IntervalUnit, Schema, TimeUnit, ToByteSlice,
+};
+use arrow::error::{ArrowError, Result};
+use arrow::record_batch::{RecordBatch, RecordBatchReader};
+use byteorder::{NetworkEndian, ReadBytesExt};
+use chrono::Timelike;
+use postgres::types::*;
+use postgres::{Client, NoTls, Row};
+
+use super::{
+    Postgres, PostgresReadIterator, EPOCH_DAYS, EPOCH_MICROS, MAGIC, UNSUPPORTED_TYPE_ERR,
+};
+use crate::SqlDataSource;
+
+impl SqlDataSource for Postgres {
+    fn get_table_schema(connection_string: &str, table_name: &str) -> Result<Schema> {
+        let (table_schema, table_name) = if table_name.contains('.') {
+            let split = table_name.split('.').collect::<Vec<&str>>();
+            if split.len() != 2 {
+                return Err(ArrowError::IoError(
+                    "table name must have schema and table name only, or just table name"
+                        .to_string(),
+                ));
+            }
+            (
+                format!("table_schema = '{}'", split[0]),
+                format!(" and table_name = '{}'", split[1]),
+            )
+        } else {
+            (String::new(), format!("table_name = '{}'", table_name))
+        };
+        let mut client = Client::connect(connection_string, NoTls).unwrap();
+        let results = client
+            .query(format!("select column_name, ordinal_position, is_nullable, data_type, character_maximum_length, numeric_precision, datetime_precision from information_schema.columns where {}{}", table_schema, table_name).as_str(), &[])
```

Comment:
If this query fails, say due to permissions, we could fall back to reading a single row, and manually building up a schema